File: leetcode-go/Q051-Q100/Q061_Rotate_List/Q061_Rotate_List.go
package Q061_Rotate_List

type ListNode struct {
	Val  int
	Next *ListNode
}

func rotateRight(head *ListNode, k int) *ListNode {
	if head == nil || head.Next == nil || k == 0 {
		return head
	}

	len := 1
	tail := head
	for tail.Next != nil {
		tail = tail.Next
		len++
	}

	// form a circle
	tail.Next = head

	for i := 0; i < len-k%len; i++ {
		tail = tail.Next
	}
	head = tail.Next
	tail.Next = nil
	return head
}

File: src/linked_list/Q061_Rotate_List.java
package linked_list;

public class Q061_Rotate_List {

    public ListNode rotateRight(ListNode head, int k) {
        int len = length(head);
        k %= len;
        ListNode runner = head;
        ListNode walker = head;
        while (k-- > 0) {
            runner = runner.next;
        }
        while (runner.next != null) {
            walker = walker.next;
            runner = runner.next;
        }
        runner.next = head;
        ListNode result = walker.next;
        walker.next = null;
        return result;
    }

    private int length(ListNode head) {
        if (head == null) {
            return 0;
        }
        return 1 + length(head.next);
    }

    class ListNode {
        int val;
        ListNode next;

        ListNode(int x) {
            val = x;
        }
    }

}

